src/store/data-store-level.ts

```diff
@@ -85,22 +111,47 @@ export class DataStoreLevel implements DataStore {
     return readableStream;
   }
 
-  public async has(tenant: string, recordId: string, dataCid: string): Promise<boolean> {
-    const partition = await this.blockstore.partition(tenant);
+  public async associate(tenant: string, messageCid: string, dataCid: string): Promise<boolean> {
+    const dataPartition = await this.blockstore.partition(DATA_PARTITION);
+
+    const exists = await dataPartition.has(dataCid);
+    if (!exists) {
+      return false;
+    }
 
-    const cid = CID.parse(dataCid);
-    const rootBlockBytes = await partition.get(cid);
+    const messagesForData = await this.blockstore.partition(HOST_PARTITION);
+    const tenantsForMessage = await messagesForData.partition(dataCid);
+    const tenants = await tenantsForMessage.partition(messageCid);
 
-    return (rootBlockBytes !== undefined);
+    await tenants.put(tenant, PLACEHOLDER_VALUE);
+
+    return true;
   }
 
-  public async delete(tenant: string, recordId: string, dataCid: string): Promise<void> {
-    const partition = await this.blockstore.partition(tenant);
+  public async delete(tenant: string, messageCid: string, dataCid: string): Promise<void> {
+    const messagesForData = await this.blockstore.partition(HOST_PARTITION);
```

Comment:
FWIW this is always a possibility with an `AbortSignal`, so not sure there's really much we can do if we allow for things to be cancelled midway

Reply:
also wouldn't it be theoretically possible for the client to send another `RecordsWrite` to clear up any data if something bad happened while trying to handle the first/previous `RecordsWrite`?

Reply:
>  possible for the client to send another RecordsWrite

Did you mean `RecordsDelete`? If so, yes, they can submit another one to fix the linger data not being deleted. But the problem is that they probably wouldn't know there is a lingering piece of data, because it would be essentially in a "detatched state" if I understand correctly.

> this is always a possibility with an `AbortSignal`

Haven't thought deeply about this, but if we didn't bother with reference counting and go with the dumb data duplication per tenant approach, we might be able to perform an atomic delete at this datastore level. Orthogonally this might also make it simple to count total storage usage per tenant.

Again not really asking you to change the implementation, just thought it's worthwhile calling out limitation so we know what's up if/when we see strange behaviors relating to this.

Reply:
oop yes i meant `RecordsDelete`

yeah fair point it would require things to be perfect in order to know that something went wrong

ill create an issue to look into ways of making storage stuff atomic (or at least safer)

Reply:
created <https://github.com/TBD54566975/dwn-sdk-js/issues/267>